### DISCUSSION TITLE
Downloaded image quality is low

### DISCUSSION THREAD

**Discussion by strtech111** [fabricjs/fabric.js#10085] [Q&A]
i am making a web phot editor 
when i download image its dimenions are same as original image but the downloaded image is of lower quality little blur 
plz see code and recommend how should i handle canvas size and downloaded image size and quality

my code is 

'use client';

import React, {useEffect, useState} from 'react';
import {fabric} from 'fabric';
import {Button, Checkbox, Input} from "@nextui-org/react";
import {MyCard} from "../../components/mycard";
import {handleFileChange} from "../../components/utils/helperFunctions";
import {OptionsTab} from "../../components/editor-components/optionsTab";
import LineColorOptions from "../../components/editor-components/LineColorOptions";
import {colorOptions} from "../../components/editor-components/colorOptions";
import ColorPicker from "../../components/ColorPicker";

const CanvasPage = () => {
    const [fabricCanvas, setFabricCanvas] = useState(null);
    const [imageUrl, setImageUrl] = useState("");
    const [selectedFiles, setSelectedFiles] = useState([]);
    const [originalImageUrl, setOriginalImageUrl] = useState("");
    const [imageDimensions, setImageDimensions] = useState({width: 0, height: 0});

    const [fillColor, setFillColor] = useState("#d51a1a"); // Default color
    const [bgColor, setBgColor] = useState("#d51a1a"); // Default color

    const [isBackground, setIsBackground] = useState(false);
    const [isBorder, setIsBorder] = useState(false);

    const [borderColor, setBorderColor] = useState("#000000"); // Default border color (black)
    const [borderWidth, setBorderWidth] = useState(1); // Default border width

    // Maximum dimensions for canvas (adjust as needed)
    const MAX_WIDTH = 800;
    const MAX_HEIGHT = 600;

    useEffect(() => {
        const canvas = new fabric.Canvas('myCanvas', {
            selection: false, // Disable selection for the entire canvas
            backgroundColor: "red"
        });

        // Set the fabric canvas in state
        setFabricCanvas(canvas);

        // Enable zooming
        canvas.on('mouse:wheel', function (opt) {
            const delta = opt.e.deltaY;
            let zoom = canvas.getZoom();
            zoom *= 0.999 ** delta;
            if (zoom > 20) zoom = 20;
            if (zoom < 0.01) zoom = 0.01;
            canvas.zoomToPoint({x: opt.e.offsetX, y: opt.e.offsetY}, zoom);
            opt.e.preventDefault();
            opt.e.stopPropagation();
        });

        return () => {
            canvas.dispose();
        };
    }, [imageUrl]);

    useEffect(() => {
        if (fabricCanvas && imageUrl) {
            fabric.Image.fromURL(imageUrl, img => {
                // Calculate the aspect ratio of the image
                const aspectRatio = img.width / img.height;

                // Determine the canvas size based on the aspect ratio
                let canvasWidth = MAX_WIDTH;
                let canvasHeight = MAX_WIDTH / aspectRatio;

                if (canvasHeight > MAX_HEIGHT) {
                    canvasHeight = MAX_HEIGHT;
                    canvasWidth = MAX_HEIGHT * aspectRatio;
                }

                // Store the original image dimensions
                setImageDimensions({width: img.width, height: img.height});

                // Update the canvas size
                fabricCanvas.setWidth(canvasWidth);
                fabricCanvas.setHeight(canvasHeight);

                img.set({
                    selectable: false, // Prevent selection
                    hasControls: false, // Disable control handles
                    lockMovementX: true, // Prevent movement on X-axis
                    lockMovementY: true, // Prevent movement on Y-axis
                    lockRotation: true // Prevent rotation
                });

                // Set the canvas background image
                fabricCanvas.setBackgroundImage(img, fabricCanvas.renderAll.bind(fabricCanvas), {
                    scaleX: fabricCanvas.width / img.width,
                    scaleY: fabricCanvas.height / img.height,
                    imageSmoothingEnabled: false
                });

            });
        }
    }, [fabricCanvas]);

    const deleteSelected = () => {
        if (fabricCanvas) {
            const activeObject = fabricCanvas.getActiveObject(); // Get the selected object
            if (activeObject) {
                fabricCanvas.remove(activeObject); // Remove the selected object
                fabricCanvas.renderAll(); // Re-render the canvas
            }
        }
    };

    useEffect(() => {
        if (fabricCanvas) {
            const activeObject = fabricCanvas.getActiveObject(); // Get the selected object
            if (activeObject) {
                activeObject.set('fill', fillColor);
                fabricCanvas.renderAll(); // Re-render the canvas
            }
        }
    }, [fillColor])

    useEffect(() => {
        if (fabricCanvas) {
            const activeObject = fabricCanvas.getActiveObject(); // Get the selected object
            if (activeObject) {
                activeObject.set('backgroundColor', bgColor); // Set the background color of the selected object
                fabricCanvas.renderAll(); // Re-render the canvas to reflect the change
            }
        }
    }, [bgColor, fabricCanvas]);

    useEffect(() => {
        if (fabricCanvas) {
            const activeObject = fabricCanvas.getActiveObject();
            if (activeObject) {
                activeObject.set({
                    stroke: borderColor,
                });
                fabricCanvas.renderAll(); // Re-render the canvas
            }
        }
    }, [borderColor]);

    useEffect(() => {
        if (fabricCanvas) {
            const activeObject = fabricCanvas.getActiveObject();
            if (activeObject) {
                activeObject.set({
                    strokeWidth: borderWidth,
                });
                fabricCanvas.renderAll(); // Re-render the canvas
            }
        }
    }, [borderWidth]);

    const handleFileChangeWrapper = (e) => {
        handleFileChange(e, ['.jpg', '.jpeg', '.png', '.gif'], setSelectedFiles);
        if (e.target.files.length > 0) {
            const newImageUrl = URL.createObjectURL(e.target.files[0]);
            setImageUrl(newImageUrl);
            setOriginalImageUrl(newImageUrl); // Store original image URL for downloading
        }
    };

    const handleDownload = () => {
        if (fabricCanvas) {
            // Save the current zoom level and viewport transform
            const currentZoom = fabricCanvas.getZoom();
            const viewportTransform = fabricCanvas.viewportTransform.slice();

            fabricCanvas.setZoom(1);
            fabricCanvas.setViewportTransform([1, 0, 0, 1, 0, 0]);

            // Create a temporary canvas for the download
            const tempCanvas = document.createElement('canvas');
            const tempCtx = tempCanvas.getContext('2d');

            // Set the dimensions of the temporary canvas to the original image size
            tempCanvas.width = imageDimensions.width;
            tempCanvas.height = imageDimensions.height;

            // Clear the temporary canvas
            tempCtx.clearRect(0, 0, tempCanvas.width, tempCanvas.height);

            // Draw the Fabric.js canvas content onto the temporary canvas
            const fabricDataUrl = fabricCanvas.toDataURL({
                format: 'png',
                quality: 1.0
            });
            const fabricImage = new Image();
            fabricImage.src = fabricDataUrl;

            fabricImage.onload = () => {
                // Draw the image on the temporary canvas
                tempCtx.drawImage(fabricImage, 0, 0, imageDimensions.width, imageDimensions.height);

                // Generate the image URL from the temporary canvas
                const dataUrl = tempCanvas.toDataURL({
                    format: 'png',
                    quality: 1.0
                });

                // Create a link element to download the image
                const a = document.createElement('a');
                a.href = dataUrl;
                a.download = 'edited_image.png'; // Name of the downloaded file
                a.click();

                // Restore the zoom level and viewport transform of the original canvas
                fabricCanvas.setZoom(currentZoom);
                fabricCanvas.viewportTransform = viewportTransform;
                fabricCanvas.requestRenderAll();
            };
        }
    };

    return (

        <div className="w-full text-center mt-4">
            <div>
                {!imageUrl && (
                    <div className="pb-0">
                        <MyCard
                            title1="Photo Editor"
                            description1="Add text, shapes, and effects to your photos."
                            accept={['.jpg', '.jpeg', '.png', '.gif']}
                            handleFileChange={handleFileChangeWrapper}
                            selectedFiles={selectedFiles}
                        />
                    </div>
                )}

                {imageUrl && (
                    <div className="grid grid-cols-5 gap-4 w-full min-w-full workarea bg-cover bg-no-repeat bg-center">

                        <div className="col-span-1">
                            <OptionsTab fabricCanvas={fabricCanvas}/>
                        </div>

                        <div className="col-span-3 flex justify-center">
                            <canvas id="myCanvas" className="w-full h-full"></canvas>
                        </div>

                        <div className="grid grid-cols-1 w-48 gap-4 max-h-96 px-3 overflow-y-auto">
                            <Button onClick={handleDownload} className="mb-4">Download</Button>
                            <Button onClick={deleteSelected}>Delete Selected</Button>

                            {/* FillColor of Selected object */}
                            {/* Color Dropdown */}
                            <label htmlFor="color-dropdown">Select Color:</label>
                            <LineColorOptions
                                id="color-dropdown"
                                lineColor={fillColor}
                                colorOptions={colorOptions}
                                handleLineColorChange={(newColor) => setFillColor(newColor)} // Pass function reference
                            />

                            <ColorPicker color={fillColor} onChangeColor={(newColor) => setFillColor(newColor)}/>

                            <Checkbox isSelected={isBackground} onValueChange={setIsBackground}>
                                Background
                            </Checkbox>

                            {isBackground && (
                                <>
                                    {/* Background of Selected object */}
                                    {/* Color Dropdown */}
                                    <label htmlFor="color-dropdown">Select BgColor:</label>
                                    <LineColorOptions
                                        id="color-dropdown"
                                        lineColor={bgColor}
                                        colorOptions={colorOptions}
                                        handleLineColorChange={(newColor) => setBgColor(newColor)} // Pass function reference
                                    />

                                    <ColorPicker color={bgColor} onChangeColor={(newColor) => setBgColor(newColor)}/>

                                </>
                            )}

                            <Checkbox isSelected={isBorder} onValueChange={setIsBorder}>
                                Stroke
                            </Checkbox>

                            {/* Border Color */}
                            {isBorder && (
                                <>
                                    <label htmlFor="border-color-dropdown">Border Color:</label>
                                    <LineColorOptions
                                        id="border-color-dropdown"
                                        lineColor={borderColor}
                                        colorOptions={colorOptions}
                                        handleLineColorChange={setBorderColor}
                                    />
                                    <ColorPicker color={borderColor} onChangeColor={setBorderColor}/>

                                    {/* Border Width */}
                                    <Input
                                        className='w-32'
                                        type="number"
                                        label="Border width"
                                        value={borderWidth}
                                        onChange={(e) => setBorderWidth(parseInt(e.target.value, 10))}
                                    />

                                </>
                            )}
                        </div>
                    </div>
                )}
            </div>

        </div>
    );
};

export default CanvasPage;

**Comment by asturur** [comment]
Without asset to test and reproducible use case i don't think anyone is going to verify this.
You should post a blob of code and ask to look at super generic code describe the problem in a generic way
